VideoLocker/src/main/java/org/edx/mobile/view/CourseBaseActivity.java

```diff
@@ -119,22 +124,57 @@ protected void onPause() {
     @Override
     protected void onDestroy() {
         super.onDestroy();
+        if (getHierarchyTask != null) {
+            getHierarchyTask.cancel(true);
+            getHierarchyTask = null;
+        }
     }
 
 
     @Override
     public void onSaveInstanceState(Bundle outState) {
-        if ( courseData != null)
-            outState.putSerializable(Router.EXTRA_ENROLLMENT, courseData);
-        if ( courseComponentId != null )
-            outState.putString(Router.EXTRA_COURSE_COMPONENT_ID, courseComponentId);
         super.onSaveInstanceState(outState);
+        outState.putSerializable(Router.EXTRA_ENROLLMENT, courseData);
+        outState.putString(Router.EXTRA_COURSE_COMPONENT_ID, courseComponentId);
     }
 
     protected void restore(Bundle savedInstanceState) {
-        if (savedInstanceState != null) {
-            courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
-            courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
+        courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
+        courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
+
+        if (courseComponentId == null) {
+            getHierarchyTask = new GetCourseStructureTask(this, courseData.getCourse().getId()) {
+                @Override
+                public void onSuccess(CourseComponent courseComponent) {
+                    if (courseComponent != null) {
+                        courseComponentId = courseComponent.getId();
+                        // The getHierarchyTask variable is set to null in onDestroy(), so this,
+                        // check ensures that the Activity has not been destroyed yet.
+                        if (getHierarchyTask != null) {
```

Comment:
Is it safer to just have a ``destroyed`` flag? That way if the task logic gets refactored at all, no one has to ensure the contract from this comment is still valid? Or is this too linked to that logic anyway?